Sources/Diff/FullDiff.php

```diff
@@ -0,0 +1,1505 @@
+							}
+							break;
+
+						case 'ins':
+							foreach ($blob['lines'] as $l => $line) {
+								$output[] = '+ ' . $line;
+								$this->formatIncompleteLines($output, $line, $h, $l, $last_h, $last_l);
+							}
+							break;
+
+						case 'replace_ins':
+							foreach ($blob['lines'] as $l => $line) {
+								$output[] = '! ' . $line;
+								$this->formatIncompleteLines($output, $line, $h, $l, $last_h, $last_l);
+							}
+							break;
+					}
+				}
+			}
+		}
+
+		return implode('', $output);
+	}
+
+	/***********************
+	 * Public static methods
+	 ***********************/
+
+	/**
+	 * Parses raw diff data to make one or more instances of this class.
+	 *
+	 * Supports raw diffs in unified format, context format, and normal format.
+	 * Does not support ed script format.
+	 *
+	 * If the raw diff contains info about multiple files, the returned array
+	 * will contain multiple instances of this class.
+	 *
+	 * In addition to the standard diff headers for the unified and context
+	 * formats, the following extended headers are also recognized:
+	 *
+	 *    optional
+	 *    rename from <old path>
+	 *    rename to <new path>
+	 *    copy from <old path>
+	 *    copy to <new path>
+	 *
+	 * The "optional" header can be manually inserted into the raw diff in order
+	 * to indicate that the changes to a particular file can be ignored if they
+	 * cannot be applied. This header should be inserted before the standard
+	 * diff headers and, where applicable, before the "rename" and "copy"
+	 * extended headers for the file that it refers to. If a mod wants to make
+	 * both mandatory and optional changes in the same file, two diffs for the
+	 * file can be included: first a diff containing the mandatory changes to
+	 * the file, and then a second diff with the "optional" header containing
+	 * the optional changes.
+	 *
+	 * The "rename" and "copy" headers are generated by `git diff` when a file
+	 * is copied or renamed. They can also be manually inserted into the raw
+	 * diff by mod authors. They must always occur in pairs with the "from"
+	 * header line immediately followed by the corresponding "to" header line.
+	 * This information is used to indicate the correct file operations to
+	 * perform when applying the diff to files.
+	 *
+	 * @see \SMF\PackageManager\PackageUtils::parseDiff() for more information
+	 *    on how to generate raw diffs for use in SMF's package manager.
+	 *
+	 * @param string $raw_diff The raw diff.
+	 * @throws \ValueError if $raw_diff does not contain valid raw diff data.
+	 * @return array Instances of this class.
+	 */
+	public static function createFromRaw(string $raw_diff): array
+	{
+		$diffs = [];
+
+		$optional = false;
+		$path_changes = [
+			'copy' => [],
+			'rename' => [],
+		];
+
+		$lines = preg_split('/\R\K/u', $raw_diff);
+
+		while (!empty($lines)) {
+			// Unified diff.
+			if (
+				str_starts_with($lines[0], '--- ')
```

Comment:
FYI, str_starts_with is a little slower than some other methods
https://gist.github.com/jdarwood007/1a949424f8cf85ca1b5f66ce38527eb6

Not that this is logic that should be heavily used all the time during the forum operations.  Just a note that we can get some better performance, which may help with larger diffs.

Reply:
Meh. I'm not really concerned about micro-optimizations for this. 🙂

Reply:
That's fine, I opened up an issue with PHP on it because in reality they should be outperforming any userland functions.  It seems a variety of issues could cause str_starts_with to be slower.